### ISSUE TITLE
virtual-scroll: support items of unknown size

### ISSUE THREAD

**Issue by mmalerba** [angular/material2#10113]
Support virtual scrolling over a list of items whose size is not known and needs to be measured

**Comment by amcdnl** [comment]
Been testing autosize and seems to work pretty well.

**Comment by fxck** [comment]
Is this released? Or are any extra steps needed? I'm getting
```
ERROR Error: StaticInjectorError(AppModule)[CdkVirtualScrollViewport -> InjectionToken VIRTUAL_SCROLL_STRATEGY]: 
  StaticInjectorError(Platform: core)[CdkVirtualScrollViewport -> InjectionToken VIRTUAL_SCROLL_STRATEGY]: 
```

**Comment by webcat12345** [comment]
Looking forward to see this feature! :+1:

**Comment by fxck** [comment]
It seems to be working fine in most of the time, except that there is always an empty space on the bottom that keeps increasing in size each time you scroll down and up a bit.

**Comment by webcat12345** [comment]
@fxck could you please share some of your code to show how to use this feature?

**Comment by fxck** [comment]
Just import `import { ScrollingModule } from '@angular/cdk-experimental/scrolling';` and use it with `autosize` directive.

```
<cdk-virtual-scroll-viewport [style.height.px]="height" autosize>
  <div *cdkVirtualFor="let item of list">{{ item }}</div>
</cdk-virtual-scroll-viewport>
```

**Comment by webcat12345** [comment]
```
  <cdk-virtual-scroll-viewport class="viewport babelfish-content" autosize>
    <bf-chat-message *cdkVirtualFor="let message of messages; let i = index"
                     [attr.id]="message?.id"
                     [message]="message"
                     [group]="currentGroup"
                     [me]="me"
                     (imageLoaded)="scrollToBottom()"
                     fxFlex="0 0 auto" fxLayout="column"></bf-chat-message>
  </cdk-virtual-scroll-viewport>```

It does not work for me, screencast attached.

https://youtu.be/sr_focoBfDM

Can you please share your thought if you have some time?

**Comment by webcat12345** [comment]
yes, it works for me :)

**Comment by fxck** [comment]
That's most likely the reason why this is experimental and not in stable. It's just not finished and feature complete yet.

**Comment by webcat12345** [comment]
I see. I have tested `autosize` feature with a list of 121000 + different height items and it works well without any lag. :+1: 
Is there any ETA for this feature? (but with normal functions as well :) ) This is really cool. We are going to use this for our project and really want to see this to be released!!!

**Comment by webcat12345** [comment]
any update on this feature?

**Comment by bop10** [comment]
@mmalerba Any updates on this?  Thanks

**Comment by skatestyle** [comment]
Will this be implemented in production soon?

**Comment by salsadoka** [comment]
I've been trying to get autosize strategy to work based on fxck's example, but keep getting this error: "Can't bind to 'cdkVirtualForOf' since it isn't a known property of 'div'. " Any suggestions?

**Comment by Anavaras** [comment]
> I've been trying to get autosize strategy to work based on fxck's example, but keep getting this error: "Can't bind to 'cdkVirtualForOf' since it isn't a known property of 'div'. " Any suggestions?

Try with the cdk version angular 6.0.2

**Comment by easyproger** [comment]
any updates ? 7.0.4 angular say - Can't bind to 'cdkVirtualForOf'  
angular 6.4.7 bug with autosize ( end scroll with space )

**Comment by jonnyleeharris** [comment]
> any updates ? 7.0.4 angular say - Can't bind to 'cdkVirtualForOf'
> angular 6.4.7 bug with autosize ( end scroll with space )

Have you tried importing both the core and experimental scroll modules? I found this aleviated the `Can't bind to cdkVirtualForOf` issue.

```
import { ScrollingModule } from '@angular/cdk/scrolling';
import { ScrollingModule as ExperimentalScrollingModule} from '@angular/cdk-experimental/scrolling';

@NgModule({
  declarations: [
    AppComponent  
  ],
  imports: [
    BrowserModule,  
    ScrollingModule,
    ExperimentalScrollingModule,    
  ],
  bootstrap: [AppComponent]
})
export class AppModule { }
```

**Comment by jonnyleeharris** [comment]
> But if I use `autosize` strategy it can not use many functions such as `scrollToIndex()` ...
> Is there any alternative way? For example I am going to scroll to bottom

I am scrolling to the bottom as follows:

```
    let top = this.viewport.measureScrollOffset("top");
    let bottom = this.viewport.measureScrollOffset("bottom");
    let offset = top + bottom;    
    this.viewport.scrollToOffset(offset);
```

However, there are a few layout bugs it appears - but it's to be expected since it is still an experimental feature.

**Comment by LeonadoRivaldo** [comment]
Any updates on this?

**Comment by easyproger** [comment]
> > any updates ? 7.0.4 angular say - Can't bind to 'cdkVirtualForOf'
> > angular 6.4.7 bug with autosize ( end scroll with space )
> 
> Have you tried importing both the core and experimental scroll modules? I found this alleviated the `Can't bind to cdkVirtualForOf` issue.
> 
> ```
> import { ScrollingModule } from '@angular/cdk/scrolling';
> import { ScrollingModule as ExperimentalScrollingModule} from '@angular/cdk-experimental/scrolling';
> 
> @NgModule({
>   declarations: [
>     AppComponent  
>   ],
>   imports: [
>     BrowserModule,  
>     ScrollingModule,
>     ExperimentalScrollingModule,    
>   ],
>   bootstrap: [AppComponent]
> })
> export class AppModule { }
> ```

yeah but im get error - itemSize need to be set... if im set 0, ofc scroll load all items... if set any size scroll work not like autosized (((

**Comment by pegak** [comment]
I have this:
```
"@angular/cdk": "^7.3.3",
"@angular/cdk-experimental": "7.3.3",
```

```
import { ScrollingModule } from '@angular/cdk-experimental/scrolling';
import { ScrollDispatchModule } from '@angular/cdk/scrolling';
```

`imports: [ScrollDispatchModule, ScrollingModule],`

And everything works ok with autosize.

**Comment by Fleeck** [comment]
@easyproger 
What worked for me is `itemSize="50"` for the `cdk-virtual-scroll-viewport` and `[style.height]="'auto'"` on the `li` without any additional height styles applied. 

I am importing both
```
import { ScrollingModule } from '@angular/cdk/scrolling';
import { ScrollingModule as ExperimentalScrollingModule } from '@angular/cdk-experimental/scrolling';
```
P.S. for some reason simple `autosize` causing occasional empty space without any list items at the bottom of the viewport.

**Comment by Apokryph** [comment]
I tried to integrate mat-expansion-panels as list items. It's causing a few bugs while scrolling.
 
1) I click on one item to open it
2) I scroll down
3) Suddenly other elements are open as well even if I didn't open them manually. The scrolling behaviour is not smooth anymore.

Does anybody know a workaround for this?

**Comment by ste-rob7** [comment]
Any update on this?  Particularly the re-introduction of the scrolledIndexChange event with variable height items.  Unless anyone knows of a work-around to be able to identify when the bottom of the list has been scrolled to, so that I know to fetch more data from the back-end?

**Comment by philly-vanilly** [comment]
@mmalerba in Ionic the item height is optional, but when it is provided, the calculation after scrolling into a new item is skipped:

https://github.com/ionic-team/ionic/blob/master/angular/src/directives/virtual-scroll/virtual-scroll.ts

```
  /**
   * An optional function that maps each item within their height.
   * When this function is provides, heavy optimizations and fast path can be taked by
   * `ion-virtual-scroll` leading to massive performance improvements.
   *
   * This function allows to skip all DOM reads, which can be Doing so leads
   * to massive performance
   */
  itemHeight?: ItemHeightFn;
```

Perhaps you can implement it in a similar way for Material?

**Comment by maciekprzysowa** [comment]
> I tried to integrate mat-expansion-panels as list items. It's causing a few bugs while scrolling.
> 
> 1. I click on one item to open it
> 2. I scroll down
> 3. Suddenly other elements are open as well even if I didn't open them manually. The scrolling behaviour is not smooth anymore.
> 
> Does anybody know a workaround for this?

I would expect it have something to do with how caching works - try setting templateCacheSize: 0 on the cdkVirtualFor and test it out.

**Comment by MEthoD112** [comment]
scrolling with auto size strategy has problems with rendering now
![345](https://user-images.githubusercontent.com/23101743/56644225-f180a680-6683-11e9-8560-a827c990809b.gif)

**Comment by pauloeli** [comment]
> Any update on this? Particularly the re-introduction of the scrolledIndexChange event with variable height items. Unless anyone knows of a work-around to be able to identify when the bottom of the list has been scrolled to, so that I know to fetch more data from the back-end?

You have found any solution or work-around? Same problem here.

**Comment by alexdabast** [comment]
Also has someone else experienced this bug ?
https://angular-virtualdom-bug.stackblitz.io/
With 2 or 3 items inside the virtual scroll (depend on the height of your item) if you scroll fast to the top (once you are at the bottom) one element disappear

**Comment by SvenBudak** [comment]
Any news here? This is ne of my most wanted features.

**Comment by mmalerba** [comment]
I'm going to be on vacation for most of July, but I've increased the priority as its something I'd like to work on after I get back

**Comment by lppedd** [comment]
Will this strategy support dynamic height of the viewport too?

**Comment by samuelblattner** [comment]
This is great! Are there any plans to integrate this with Material Flat Trees directly? I have documents with large tree structures (with leaves of different sizes) that I would like to display side by side. Virtual scrolling seems to work pretty well when rendering the flattened tree structure manually within the viewport. It would be awesome if I cloud just activate virtual scrolling with a setting in FlatTreeControl or so...

**Comment by SvenBudak** [comment]
I check every day about 10 times this issue, cant wait to get this feature :D

**Comment by Frotty** [comment]
Hello, we are also looking forward to this feature.
The experimental version is working almost correctly for us with but 2 flaws:
1. When initially loading a virtual scroll with many elements, but 2 rows stretching across the entire view area, more than two rows are loaded until you scroll to the last element.
2. After scrolling to the last element, it is impossible to scroll all the way back up. The view is cut off.

**Comment by cesco69** [comment]
@mmalerba any news?

**Comment by prashanthc** [comment]
We need this feature. Still waiting for it. :smile_cat:

**Comment by sufiiiyan** [comment]
> It seems to be working fine in most of the time, except that there is always an empty space on the bottom that keeps increasing in size each time you scroll down and up a bit.

Facing the same issue

**Comment by jaigtz88** [comment]
> Any update on this? Particularly the re-introduction of the scrolledIndexChange event with variable height items. Unless anyone knows of a work-around to be able to identify when the bottom of the list has been scrolled to, so that I know to fetch more data from the back-end?

Did you resolve it?

**Comment by jsachica** [comment]
This is such a cool feature, any plans on releasing it to production yet?

**Comment by Frotty** [comment]
> This is such a cool feature, any plans on releasing it to production yet?

For that it would need to work properly, I suppose.

**Comment by SvenBudak** [comment]
Cant release my image board before this feature is done ;P

**Comment by saif003** [comment]
As of now is there any way to have responsive list design along with virtual scroll?

I'm using Angular Flex Layout along with virtual scroll, so I believe we can be aware of what will be the size of items on any specific screen size, but can we bind itemSize property to populate it dynamically,

it won't be essentially updated once assigned apart from probably when orientation changes on Tabs or cellular devices.

**Comment by SvenBudak** [comment]
someone got a good working service to calculate (without to much lost of performance) the hight of items?

**Comment by tibinthomas** [comment]
Scrolling down with autosize directive is working fine but while scrolling up its jumping to the start of the list.

**Comment by tibinthomas** [comment]
Anyone in the world other than me worrying about this? Is this issue a solvable or can we solve it by proving it can't be solved?

**Comment by yadue** [comment]
@SvenBudak @tibinthomas what I did to solve that issue instead of scrolling directly to the selected element I scroll element by it's height height to the top or bottom, check if element I look for is visible, if not scroll again until element is found. It's not perfect but it works in most situations.

**Comment by Frotty** [comment]
@mmalerba said he would get back to this after vacation, but seems like they are working on other things now :/

**Comment by Splaktar** [comment]
Yes, supporting Ivy and creating new Test Harnesses has taken priority over this work. This is still a high priority issue that the team intends to address.

**Comment by noobcoder15** [comment]
any news when this is going to be released?

cdk-virtual-scroll-viewport makes no sense to me with a fixed height.

**Comment by bojidaryovchev** [comment]
https://pastebin.com/Mqsdrucm here you go guys - today I played with that and managed to come up with this - its a custom virtual scroll strategy - an abstract generic class with a protected abstract calculateItemHeight function which you need to override in an extending class - the constructor accepts an array of items (some models) which u need to pass when overriding and the calculateItemHeight function takes one item and calculates its height (for example u have some nested items and they have some nested items) - the most important thing is that those inner items must be of fixed height so u could use some constants and multiplication to calculate their height, otherwise you cannot know their height before them being rendered, and calculating the proper height is crucial for the scroll to work 100% properly - with those remarks taken into account, it should work just fine

**Comment by dauledk** [comment]
> the most important thing is that those inner items must be of fixed height

Is this thread not virtual scroll, for items of unknown size?

**Comment by bojidaryovchev** [comment]
> Is this thread not virtual scroll, for items of unknown size?

I know its insufficient I'm just trynna by helpful in some way

**Comment by bojidaryovchev** [comment]
ok so I got a hack to get dynamic size working - using the strategy I provided above, I modified it (no longer abstract unfortunately) so the changes are:

 ```
 constructor(private readonly component: { loaded: Subject<boolean>; someModelType: someModelType[]; cdkVirtualScrollContainer: HTMLElement; entry: TemplateRef<unknown>; }) {
    component.loaded.subscribe(() => {
      for (let index = 0, offset = 0; index < this.component.someModelType.length; index++) {
        const element = this.component.someModelType[index];
        const elementHeight = this.calculateItemHeight(element, this.component.cdkVirtualScrollContainer, this.component.entry);
  
        offset += elementHeight;
        
        this.offsetByIndex[index] = offset;
        this.indexByOffset[offset] = index;
      }
    });
  }
```

and the calculateItemHeight function was modified to

  ```
private calculateItemHeight(someModelType: SomeModelType, container: HTMLElement, templateRef: TemplateRef<unknown>): number {
    const embeddedViewRef: EmbeddedViewRef<unknown> = templateRef.createEmbeddedView({ someModelType: someModelType});

    embeddedViewRef.detectChanges();

    const element = embeddedViewRef.rootNodes.shift() as HTMLElement;

    container.appendChild(element);

    const height = element.getBoundingClientRect().height;

    element.remove();

    return height;
  }
```

what is the idea - well I wrap the thing Im using *cdkVirtualFor on in a ng-template and pass that template to the strategy so it dynamically renders the items and appends it to the cdk scroll container to measure its height and then removes it - I know this is a hack but its working

**Comment by Frotty** [comment]
It's not. I wouldn't hold my breath. This library supports varying row height with a cache:
https://github.com/rintoj/ngx-virtual-scroller

**Comment by thackerronak** [comment]
in autosize, scroll down event working fine while scrolling up facing jump issue when scrolling gets to stop.  any update for the same?

**Comment by jocker** [comment]
Here's a brute approach to a very annoying missing functionality that nobody from the angular team seems to care about.

```typescript

i
interface RowInfo {
    height: number
    id: string
}

export class CustomVirtualScrollStrategy implements VirtualScrollStrategy {
    
    private viewport:CdkVirtualScrollViewport;
    private rowSizeCache = new Map<string, RowInfo>();
    private orderedItemIds:string[] = [];
    scrolledIndexChange = new Subject<number>();
    attachedSubject = new Subject<CdkVirtualScrollViewport>();


    constructor(private minItemHeight:number=48, private renderItemsBefore=2, private renderItemsAfter=2) {
    }

    onContentRendered(): void {
    }

    attach(viewport: CdkVirtualScrollViewport): void {
        if(viewport == this.viewport){
            return
        }
        this.attachedSubject.next(viewport);
        this.viewport = viewport;

        const changeObserver = new Observable<HTMLElement>(subscriber => {

            const watchedEl = viewport.elementRef.nativeElement.firstChild as HTMLElement;

            const mutationObserver = new MutationObserver(() => {
                subscriber.next(watchedEl)
            });

            mutationObserver.observe(viewport.elementRef.nativeElement.firstChild, {
                attributes: true,
                //characterData: true,
                childList: true,
                subtree: true,
                attributeOldValue: true,
                //characterDataOldValue: true
            });

           return () => {
               mutationObserver.disconnect();
           }
        });

        changeObserver.pipe(
            debounceTime(100),
            takeUntil(this.attachedSubject)
        ).subscribe(rootEl => {
            let changed = false;
            // each child of cdk-virtual-scroll-viewport should have data-row-id and data-row-index attributes defined
            // <div *cdkVirtualFor="let row of dataSource.onData() | async; let index = index;" [attr.data-row-id]="row.getId()" [attr.data-row-index]="index">
            const list = Array.from(rootEl.querySelectorAll("[data-row-id][data-row-index]")) as HTMLElement[];

            // TODO handle item removal
            // TODO handle reordering items
            // TODO various other optimizations can be performed as we don't always need to check each row height
            
            for(const el of list){
                const id = el.getAttribute('data-row-id');
                const idx = parseInt(el.getAttribute('data-row-index'), 10);
                let rowInfo = this.rowSizeCache.get(id);
                if(!rowInfo){
                    rowInfo = {height:0, id: id};
                    this.rowSizeCache.set(id, rowInfo)
                }

                const clientHeight = el.clientHeight;

                if(rowInfo.height != clientHeight){
                    rowInfo.height = clientHeight;
                    changed = true
                }

                if(!isNaN(idx)){
                    this.orderedItemIds[idx] = id;
                }
            }

            if(changed){
                this.updateTotalContentSize();
                this.updateRenderedRange()
            }

        })

    }

    private updateRenderedRange() {
        if (!this.viewport) {
            return;
        }

        const scrollOffset = this.viewport.measureScrollOffset();
        const viewportSize = this.viewport.getViewportSize();
        const maxOffset = scrollOffset+viewportSize;
        const maxLastVisibleIndex = Math.ceil(maxOffset / this.minItemHeight);

        const dataLength = this.viewport.getDataLength();

        let offset = 0;
        const visibleRange = {startIndex:NaN, endIndex:NaN, startOffset:0, endOffset:0}; // contains rowSizeCache which are partially or fully visible
        for(let i=0;i < Math.min(dataLength, maxLastVisibleIndex); i++){

            let itemHeight = this.minItemHeight;
            const itemId = this.orderedItemIds[i];
            if(itemId){
                itemHeight = this.rowSizeCache.get(itemId).height
            }

            if(offset > maxOffset){
                break;
            }

            if(offset+itemHeight >= scrollOffset){
                if(isNaN(visibleRange.startIndex)){
                    visibleRange.startIndex = i;
                    visibleRange.startOffset = offset;
                }

                visibleRange.endIndex = i;
                visibleRange.endOffset = offset + itemHeight;
            }

            offset += itemHeight;

        }

        const firstVisibleIndex = visibleRange.startIndex;

        for(let i=this.renderItemsBefore; i>0 && visibleRange.startIndex > 0; i-=1){
            visibleRange.startIndex -= 1;
            visibleRange.startOffset -= this.getItemHeight(visibleRange.startIndex)
        }

        for(let i=this.renderItemsAfter; i>0 && visibleRange.endIndex < dataLength-1; i-=1){
            visibleRange.endIndex += 1;
            visibleRange.endOffset += this.getItemHeight(visibleRange.endIndex)
        }

        this.viewport.setRenderedRange({start:visibleRange.startIndex, end: visibleRange.endIndex+1});
        this.viewport.setRenderedContentOffset(visibleRange.startOffset);
        this.scrolledIndexChange.next(firstVisibleIndex);
    }

    private updateTotalContentSize() {
        if (!this.viewport) {
            return;
        }

        let totalHeight = 0;

        for(let i=0; i< this.viewport.getDataLength(); i++){
            totalHeight += this.getItemHeight(i);
        }

        this.viewport.setTotalContentSize(totalHeight);
    }

    private getItemHeight(index: number): number{
        const itemId = this.orderedItemIds[index];
        if(itemId){
            return this.rowSizeCache.get(itemId).height
        }
        return this.minItemHeight;
    }


    detach(): void {
        if(this.viewport){
            this.viewport = null;
            this.attachedSubject.next(null)
        }
    }

    onContentScrolled(): void {
        this.updateRenderedRange();
    }

    onDataLengthChanged(): void {
        this.updateTotalContentSize();
        this.updateRenderedRange();
    }

    onRenderedOffsetChanged(): void {
    }

    scrollToIndex(index: number, behavior: "auto" | "smooth"): void {
        //TODO
    }


}


```

**Comment by dudipsh** [comment]
Working Example 

#### feed.component.html
```html
 <cdk-virtual-scroll-viewport
                  #viewport class="example-viewport" [style.height.px]="getHeight"
                                             autosize fxFlex>
                    <div  #listComponent  fxFlex fxLayout="column" *ngIf="messages && messages.length > 0">
                        <div
                             *cdkVirtualFor="let item of messages;let index = index; templateCacheSize: 0;
                                trackBy: indexTrackFn; let i = index"
                                  class="feed-messages"  [style.height.px]="item">
                            <app-message *ngIf="item?.senderId"
                                         [messageInput]="item"
                                         [messageUser]="chatUsers.users[item.senderId]"
                            ></app-message>
                        </div>
                    </div>
                </cdk-virtual-scroll-viewport>
```
#### feed.component.ts
```typescript
export class FeedComponent implements OnInit, AfterViewInit , OnDestroy {
  @ViewChild(CdkVirtualScrollViewport, {static: true}) viewport: CdkVirtualScrollViewport;
  @Input('isPageLoaded') set setIsPageLoaded(isPageLoaded) {
    if (isPageLoaded) {
      (<any>this.viewport).checkViewportSize();
      console.log('isPageLoaded', isPageLoaded);
    }
  };
ngOnInit(): void {
    fromEvent(window, 'resize')
      .pipe(
        distinctUntilChanged(),
        debounceTime(10),
        takeUntil(this.deactivatedSubject)
      ).subscribe(() => {
           (<any>this.viewport).checkViewportSize();

    });
    this.viewport.elementScrolled().subscribe(() => {
      console.log(this.viewport.getRenderedRange()) //{start: 4, end: 19}
    })
  }

}
```

**Comment by arkadiusz-wieczorek** [comment]
@dudipsh Can you share online example? 

I'm trying with 9.2.0 version and receiving error message:
```
ERROR Error: "Error: cdk-virtual-scroll-viewport requires the "itemSize" property to be set."
```

**Comment by dudipsh** [comment]
@arkadiusz-wieczorek 

did you import the ScrollingModule from cdk-experimental?


```typescript
import { ScrollingModule } from '@angular/cdk/scrolling';
import { ScrollingModule as ExperimentalScrollingModule} from '@angular/cdk-experimental/scrolling';

@NgModule({
  declarations: [
  ],
  imports: [
    ScrollingModule,
    ExperimentalScrollingModule,
  ]
})
export class FeedModule { }
```

**Comment by arkadiusz-wieczorek** [comment]
thanks @dudipsh, my fault!

**Comment by pjc2007** [comment]
Looking at using an **expander** in virtual scroll ie a click a button and an item height expands to contain a sub-menu, which can vary per item.

Is the plan for this this be able to also support this?

**Comment by ewalddieser** [comment]
Autosize works but there are other features missing then. For example scrolledIndexChange is not working. But I really need to know which element is at the top.

**Comment by buenjybar** [comment]
> Autosize works but there are other features missing then. For example scrolledIndexChange is not working. But I really need to know which element is at the top.

Have you tried to use the viewport to get that information?
using the `this.viewport.getRenderedRange()`

see the example : https://github.com/angular/components/issues/10113#issuecomment-607102447

**Comment by dudipsh** [comment]
@buenjybar 

```typescript
    ngOnInit(): void {
        this.route.paramMap.subscribe((res) => {
            this.activeChannel = res.get('id');
            this.chatService.readMessagesInChannel(this.activeChannel, 0, 40).subscribe((channelData: any) => {
                this.messages = channelData.data.reverse().map((item, index) => {
                    this.scrollTo(index * 299, 'auto');
                    return item;
                });
                this.startIndex = channelData.totalCount;
            });
        });

        this.viewport.elementScrolled()
            .pipe(debounceTime(150))
            .subscribe((res) => {
                const {start, end} = this.viewport.getRenderedRange();
                if (start === 0) {
                    this.fetchMore();
                }
            });

    }

    scrollTo(size, behavior: 'auto' | 'smooth' = 'auto') {
        setTimeout(() => {
            this.viewport.elementRef.nativeElement.scrollTo({top: size, behavior, left: 0});
        }, 300);
    }
```

**Comment by ewalddieser** [comment]
> > Autosize works but there are other features missing then. For example scrolledIndexChange is not working. But I really need to know which element is at the top.
> 
> Have you tried to use the viewport to get that information?
> using the `this.viewport.getRenderedRange()`
> 
> see the example : [#10113 (comment)](https://github.com/angular/components/issues/10113#issuecomment-607102447)

Yes, sure with some hack you can make it work. It still would be nice if the API that the components provide would work no matter if item size is fixed or dynamic. I mean the (scrolledIndexChange) output.

**Comment by dudipsh** [comment]
I usually agree with this approach
But this issue opened from 2018, it looks like it's time for "hacks"

By the way
[Denis Hilt](https://github.com/dhilt)
works on a solution for a chat experience
virtual scroll that starts from the bottom with auto height
its in progress but its looks like a very good solution
[https://github.com/dhilt/ngx-ui-scroll](https://github.com/dhilt/ngx-ui-scroll)

working example 
[https://stackblitz.com/edit/ngx-ui-scroll-chat-app](https://stackblitz.com/edit/ngx-ui-scroll-chat-app)

**Comment by teland94** [comment]
When it will released? Or Angular Material team forgot about this?

**Comment by dudipsh** [comment]
> @dudipsh Can you share online example?
> 
> I'm trying with 9.2.0 version and receiving error message:
> 
> ```
> ERROR Error: "Error: cdk-virtual-scroll-viewport requires the "itemSize" property to be set."
> ```

@arkadiusz-wieczorek 
I moved to this package https://github.com/dhilt/ngx-ui-scroll no issues... work perfectly

**Comment by ahnpnl** [comment]
I am looking forward to this feature too

**Comment by vraddy** [comment]
'cdk-virtual-scroll-spacer' is not taking height beyond 22339600px, that means user can not scroll down further if list showing in virtual table is beyond 22339600px height.
example: i have log viewer with millions of records, cdk-virtual-scroll-spacer height is calculated based on itemSize * numberOfItems.
to show 2 millions of records with 20px row height need 40million px cdk-virtual-scroll-spacer height, but chrome is limiting  cdk-virtual-scroll-spacer height to 22339600px.
how do we overcome this limit ?

**Comment by Frotty** [comment]
@vraddy this has nothing to do with this ticket. Also if you think scrolling through 2 million entries is feasible, reconsider your design choices (i.e. truncate your data).

**Comment by SvenBudak** [comment]
any news about this ticket?

**Comment by Celend** [comment]
Any progress for this issue?

**Comment by thackerronak** [comment]
ngx-virtual-scroller lib is best for unknown size just used it till any next update.
https://github.com/rintoj/ngx-virtual-scroller/

**Comment by SvenBudak** [comment]
@thackerronak its not really working. the guy who created this one is not available for accepting merge requests and he also cant update the npm package. My team was testing community updates from ngx-virtual-scroller but its not working well. alot bugs and problems.

We have to wait for cdk solution. The Ionic Team wrote anything about that ion-virtual-scroll will switch to cdk. maybe ionic team is working together with the angular team on this problem. we wait already 4 years. I am not sure that this fix will come any day... i have already 4 apps they are ready to release if this cdk fix is out. i worked last 5 years on them... i just hope it will come...

**Comment by Mmatiasn** [comment]
Does https://github.com/rintoj/ngx-virtual-scroller/ work with tables that have sticky headers?

**Comment by dudipsh** [comment]
@Mmatiasn 
i used "ngx-virtual-scroller" ( for NPM package ) 
you can try to fork and implement sticky headers 
https://github.com/dudipsh/ngx-ui-table

**Comment by milanov** [comment]
Has anyone stumbled across the requirement for items of different, but known size? Like a mixture between the fixed and autosize scrolling strategies? I've elaborated more in this [SO](https://stackoverflow.com/questions/66183118/angular-virtual-scroll-strategy-for-different-fixed-size-items) question.

**Comment by ms-dosx86** [comment]
Sad to say, but the performance is much worse comparing to the fixed scrolling strategy.

**Comment by xjuanc** [comment]
3+ years so far...

**Comment by SvenBudak** [comment]
dont wait for a solution. add meanwhile a intersection observer which detechts that the user reach the end of the list and load then the next X entries. This is how we did it for now. I dont think that we will get for the web a good working virtual scroll. It seems that this is a feature that works only well in native java i think.

**Comment by xjuanc** [comment]
Yeah, I'm not really waiting for it, In the end I managed to use multiple workarounds for my usecases but it just amazes me the long time this issue has been around.

**Comment by SvenBudak** [comment]
This package also has its problems. Just like many others. After almost 2 years we have accepted that there is simply no "perfect" solution written in JS.

**Comment by y0nd0** [comment]
What is when using ngFor inside an item of the virtual scroll. Which will increase the height dynamically. In my case an item can be between 70 and 520 px. 

```html
<cdk-virtual-scroll-viewport itemSize="100">
  <ng-container *cdkVirtualFor="let item of items">
    <some-card>
      <some-list>
        <some-item>
          <some-header>{{item.name}}</some-header>
        </some-item>

        <some-list>
          <!-- Here some dynamic heights. -->
          <some-item *ngFor="let option of item.options">
            <some-text>{{option}}</some-text>
          </some-item>
        </some-list>
      </some-list>
    </some-card>
  </ng-container>
</cdk-virtual-scroll-viewport>
```

You maybe think: It's rediculous, this will never work. "No god! No god, please! Nooooo". 
But the [Ionic Virtual Scroll](https://ionicframework.com/docs/api/virtual-scroll) works pretty nice. The only problem I have is then the items array changes while I'm not on scroll top 0. Overlapping items, etc. 

The current Angular CDK Virtual Scroll does not work in my case. Unusable. Since the items / pos is jumping and flickering. Regardless of the height I choose. And I cannot test the cdk-experiemental vscroll autosize, because of Angular imports issues with meaningless errors.

Any idea or recommendations (usuage or other libs) are welcome. 🙂 
Maybe my case will help to improve that component.

**Comment by milanov** [comment]
Hi Angular team. Picking up from https://github.com/angular/components/issues/10113#issuecomment-778660378 (creating a strategy for scrolling items with different, but fixed sizes). I've modified the code suggested in the SO question and it seems to work. The usage is very close to the existing strategy: `<cdk-virtual-scroll-viewport [itemSizes]='worklistItemsDataSource.heights' class="worklist-container">`.
1) Do you think the use case is generic enough so that i try to create a PR in the experimental cdk, and eventually this becomes an official virtual scrolling strategy? If not, it can be published as a standalone npm module.
2) Is there a set of tests that are meant for different scrolling strategies, something like acceptance tests that cover basic functionality? I attempted to reuse the [ones for the fixed strategy](https://github.com/angular/components/blob/master/src/cdk/scrolling/virtual-scroll-viewport.spec.ts#L32), but they also include template cache/track by/etc tests.

**Comment by Asiabi** [comment]
Why all its github github youre think github is verry good?

**Comment by Klaster1** [comment]
@milanov the full strategy (includes viewport resize, which I decided to omit) I've adapted the SO answer from is covered with unit tests, so it's definitely possible. If you wish to create a module based on it, that'd be awesome, I would have used it instead of rolling out a custom strategy.

**Comment by mmalerba** [comment]
@milanov Rather than taking an array input like in the SO question, I think a function would be a more flexible API, e.g.:

HTML:
```html
<cdk-virtual-scroll-viewport customVirtualScrollStrategy [itemSize]="itemSizeFn">
```

TS:
```ts
itemSizeFn = (item, index) => index % 2 ? 50 : 100;
```

**Comment by milanov** [comment]
Thanks for the quick responses. Some followups for @mmalerba:
- I’m still not sure if I should attempt to create a PR in the experimental cdk repository, or publish the strategy separately
- (depending on the above) Would you expect the new strategy to have the same API as the current ones do (`min/maxBufferPx`). The implementation right now supports what number of elements are to be rendered before/after the visible part of the scroll (lets say 3 elements), which makes the calculations much simpler.
- I don't think we can have an `itemSizeFn` with a signature of `(item, index)`, as we don’t actually have access to the items in the scrolling strategy, but `index` alone would be doable

As for @Klaster1 , can you elaborate on the viewport resize part? I actually modified your code to be as close as possible to the fixed size strategy (in terms of code structure), and it seems there isn't anything missing, simply the calculations in the `_updateRenderedRange` are different.

**Comment by mmalerba** [comment]
I think we can make some kind of change to this repository, I just want to make sure we settle on a good API first. That makes sense about `itemSizeFn`, I figured there was a reason I didn't implement it already, and its probably because its not really possible without some big refactoring. What if for now we change the `itemSize` to take `number | number[]` instead of just `number`. That way we don't need to add a new strategy, we could just expand the current one.

Also, I think rather than making the user enumerate all the heights, it would be nice to accept an array like `[10, 100]` which would result in alternating heights of `10` and `100` for all of the items. (As opposed to what's shown in the stackblitz example you linked to, where it uses an array of 500 sizes to accomplish the same thing: `Array.from(Array(500).keys()).map(i => (i % 2 ? 50 : 100))`

**Comment by milanov** [comment]
A assume different people will have different use-cases, so lets hear opinions from the community about the API. From my perspective, the way to go would be:
* Introduce a new virtual scrolling strategy with an `itemSizeFn` parameter, taking the `index` of a given element and returning its height. This new strategy will be in cdk-experimental and would render a predefined number of items before/after the currently visible viewport.
* (Release and take feedback from the community)
* Add `minBufferPx` and `maxBufferPx`. This will not be a breaking change and will align this strategy with the currently existing ones.
* (Release and take feedback from the community)
* (Optional) Merge the two strategies into one that takes either `itemSize` or `itemSizeFn`, thus eliminating duplication of code/tests. This should be done only if the differences in the calculations can be properly isolated.

As a side note, I'm not a fan of the `minBufferPx` and `maxBufferPx` idea, I find it simpler to reason about the behaviour in terms of how much items should be rendered before/after the viewport. However, I assume that this API is here to stay, and therefore the suggestion above.

**Comment by zergeborg** [comment]
Could someone explain me what was wrong with jocker's brute force implementation posted earlier in this thread? Was it downvoted because of performance issues, maintainability issues, security issues, or something else?

I am asking because I also need to support a list of items of variable item size, which is very different from milanov's requirement (of different, but fixed sizes). And his solution while not very elegant, gets the job done.

So far the only working solution which also supports at the end of the day the `scrollToIndex` API was posted by jocker I think. So I am curious what's the main reason for the downvotes.

**Comment by zergeborg** [comment]
@milanov I only have a very subjective, but probably relevant feedback. It would be really nice if the name of the strategy would highlight the fact that the strategy is not an auto-size strategy. Just off the top of my head `StaticVariableSizeScrollStrategy` or `ProgrammaticVariableSizeScrollStrategy` comes to my mind as opposed to e.g. a dynamic auto size strategy.

**Comment by milanov** [comment]
@zergeborg Thanks for the feedback. With regards to naming, right now I'm leaning towards `FixedSizesVirtualScrollStrategy` (with the `s` in sizes). What I'm suggesting will not work for your case, but I'm hoping it might work for others.  The three strategies (fixed size, multiple fixed sizes, dynamic size) should be complementary to each other.

**Comment by nathanaelmayne** [comment]
It looks like the autosize is included in @angular/cdk-experimental. I noticed the paragraph under the example here https://material.angular.io/cdk/scrolling/overview#scrolling-over-fixed-size-items

**Comment by CiriousJoker** [comment]
Is there any ETA for when cdk-experimental will be stable?

Specifically:`

`ERROR Error: cdk-virtual-scroll: scrolledIndexChange is currently not supported for the autosize scroll strategy`

**Comment by juri33** [comment]
"@angular/cdk": "12.2.8", 
"@angular/cdk-experimental": "12.2.8" 

If i have set only autosize, i get an error Error: cdk-virtual-scroll-viewport requires the "itemSize" property to be set.

**Comment by sayers24** [comment]
Was "@angular/cdk-experimental" removed from v13? It is saying it cannot be found.

The need for the autosize functionality seems very common for anyone developing a feed-type experience with variable content length for each entry. Has this ask really been idle for nearly four years?

**Comment by tobiasmuecksch** [comment]
Is this feature ever going to be implemented?

**Comment by asifnkhan** [comment]
Any update on this? Is this not being considered anymore?

**Comment by gtteamamxx** [comment]
any update?

**Comment by tobiasmuecksch** [comment]
@gtteamamxx I have given up hope on this. This issue is +4 years old.

There is a pretty good react plugin. Maybe it's possible to integrate that into angular somehow.

**Comment by ewalddieser** [comment]
> @gtteamamxx I have given up hope on this. This issue is +4 years old.
> 
> There is a pretty good react plugin. Maybe it's possible to integrate that into angular somehow.

@tobiasmuecksch 

It is indeed possible to use react components in angular. Here's an article. 
https://netbasal.com/using-react-in-angular-applications-1bb907ecac91

Which React component would you recommend for this?

**Comment by tobiasmuecksch** [comment]
@ewalddieser According to this [Stackoverflow question](https://stackoverflow.com/questions/45352436/dynamic-height-with-infinite-scroll-react-virtualized) `react-virtualized` contains a component called `CellMeasurer` which should solve most of the use cases.

I haven't had the time yet, to try to integrate react components. If you are successful, I'd would be nice to share your experience here.

**Comment by SvenBudak** [comment]
If you have something like a image wall where all images have the same size, but they are depending on parent container width, 
 should it now be possible in a performant way with: https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver

Just get the size of the first image and pass it to the virtual scroll. Ofc its not the solution for people they have alot different sized elements... but maybe it helps people they are trying to make a image board like me.

**Comment by tobiasmuecksch** [comment]
@SvenBudak Thank you for your input. But as this issue's title suggests, we need a solution for items of unknown (and potentially unequal) size.

My use case is a chat app. The items of interest are chat bubbles, which by nature have individual sizes.

**Comment by kevinkishoretsi** [comment]
Is it any other possibility to implement auto size in the virtual scroll

**Comment by KanaguT** [comment]
Do we have any update on production release date for this ?

**Comment by JoranLive** [comment]
4 year later, still the same.

Dont expect something...

**Comment by aymericl** [comment]
Any update ? Someone has a a workaround ? How do you deal with this problem ?
Thanks.

**Comment by spike-rabbit** [comment]
@mmalerba we have the need for the `autosize` strategy in a project, but using experimental is something that should be avoided. I could invest some time to complete the implementation there so that `autosize` can be moved to the `@angular/cdk` package. Before I start, just a few questions:

- Is this even open for community contributions?
- What is needed to make `autosize` stable?  I already saw that things related to scrollIndex are missing
- Why are these things not implemented yet? Time or are there any major problems?
- Do you have any suggestions how the missing features can be implemented?

Sorry in case I missed something, the issue is quite long already

**Comment by mmalerba** [comment]
@spike-rabbit Yes, the reason for it being still experimental is that `scrolledIndexChange` and `scrollToIndex` are not implemented. It's been a while since I looked at it, so I don't remember the exact problem I ran into trying to implement them, but I do remember it being tricky. So to answer your questions:
- Is it open for contributions: I'm willing to review a contribution for this, but just so you're aware, despite being experimental this is used by a few teams inside Google, which means I'll need to confirm it doesn't break their apps before any solution can be merged
- What's needed to make it stable: Just the 2 methods I mentioned above + tests that confirm they work as expected. I'd probably want to land those changes in cdk-experimental, then do a followup PR to graduate it to cdk
- Why its not done yet: Mainly a lack of time / other higher priorities, but I do remember it being somewhat tricky to implement correctly
- Suggestions for implementing: Unfortunately I don't really have much context on it at this point. The general idea of how the autosize virtual scroll works is that it keeps track of the amount of error between where it thinks our theoretical scroll position should be and where it actually is and then try to slowly eliminate that error as the user scrolls. There's also some buffer elements rendered before and after the ones actually visible in the viewport, which might make it tricky figuring out which one is that actual current scrolled to index (it won't necessarily just be the first one rendered)

**Comment by timelinelink** [comment]
Hey @spike-rabbit thanks for trying to make some progress on this. Any idea if/when those changes may get merged? are you waiting on @mmalerba or something else?

**Comment by spike-rabbit** [comment]
No real idea. This is some kind of side project for me, which I work on when nothing else is to do. 

ATM I need to write proper unit tests for my changes before having a full review.  Unfortunately, I am quite busy with other stuff.

However, I definitely plan to continue working here.

**Comment by khushbupatel2** [comment]
why when I use autosize  time scrolledIndexChange and scrollToIndex are not work can you help: what should I do>

**Comment by HyperLife1119** [comment]
It would be great if we could implement virtual scrolling (e.g. @for, @defer) on the control flow syntax!

**Comment by bmwz110** [comment]
`An autosize strategy that works on elements of differing sizes is currently being developed in @angular/cdk-experimental, but it is not ready for production use yet.`
-- from Angular Material Doc

It looks like this is exactly what the developers wanted, hope to see this support in official version soon.🌹

**Comment by yinshuxun** [comment]
any update?

**Comment by xetrics** [comment]
I'd like to help move this feature along. What is the current status of what still needs done for this feature to be shippable?
@mmalerba

**Comment by lean1024** [comment]
any updates?

**Comment by Elmanderrr** [comment]
it's been 6 year since issue was created, maybe it's time?

**Comment by stephaniacc93** [comment]
Any updates?

**Comment by BearandYoon** [comment]
Any updates?

**Comment by erzenhaxhi** [comment]
7 years, 33 days, 9 hours, 2 minutes, and 16 seconds

**Comment by TidianeRolland** [comment]
Any updates ?

**Comment by rdlabo** [comment]
Last year, I created a library for the strategy of specifying sizes directly. This is a directive that is used in combination with cdk-virtual-scroll-viewport, and it is working well so far.

Will this library be helpful to anyone? I hope so.

https://github.com/rdlabo-team/ionic-angular-library/tree/main/projects/scroll-strategies#readme

DEMO: https://rdlabo-ionic-angular-library.netlify.app/main/scroll-strategies/advanced

**Comment by jsuterICBF** [comment]
The usual any update on progress? Will we get it before GTA VI?

**Comment by armando-ota** [comment]
Any updates ?

**Comment by Harpush** [comment]
I think this is a good compromise solution for now: https://github.com/angular/components/issues/33045
It allows "dynamic" size but without automatic measuring.

**Comment by yinshuxun** [comment]
+1

---- Replied Message ----
| From | Armando ***@***.***> |
| Date | 04/20/2026 13:38 |
| To | angular/components ***@***.***> |
| Cc | yinshuxun ***@***.***>,
Comment ***@***.***> |
| Subject | Re: [angular/components] virtual-scroll: support items of unknown size (#10113) |
armando-ota left a comment (angular/components#10113)

Any updates ?

**Comment by jared-ceriousdevtech** [comment]
For anyone blocked by production-ready variable heights in Angular, I am working on @ceriousdevtech/ngx-cerious-scroll as a focused alternative. It is not a CDK drop-in, but it is designed around dynamic item heights from the start.
